### PR TITLE
Add support for CPU AMP autocast

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds PyTorch 1.10 support for CPU autocast. Requires installing 1.10 (and I noticed that for now till the release, we have to include a dev check to the availability for people using nightly or building from source).

Now in 1.10 this works:

```python
Trainer(precision='bf16')
```

cc @ananthsub @anj-s @min-xu-ai I noticed that torch.cpu.autocast only has `bf16` support which means it might not play well with `fp16` mode for FSDP when using CPU offloading. Would be nice to have support for this out the box!

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/imports.py [hunk 1]
```diff
@@ -68,7 +68,9 @@ def _compare_version(package: str, op, version) -> bool:
 _TORCH_GREATER_EQUAL_1_8 = _compare_version("torch", operator.ge, "1.8.0")
 _TORCH_GREATER_EQUAL_1_8_1 = _compare_version("torch", operator.ge, "1.8.1")
 _TORCH_GREATER_EQUAL_1_9 = _compare_version("torch", operator.ge, "1.9.0")
-_TORCH_GREATER_EQUAL_1_10 = _compare_version("torch", operator.ge, "1.10.0")
+_TORCH_GREATER_EQUAL_1_10 = _compare_version("torch", operator.ge, "1.10.0") or _compare_version(
+    "torch", operator.ge, "1.10.0dev"
+)  # todo (sean): remove dev check once 1.10 released
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:73]: Perhaps you should use a separate check similar to this one:

https://github.com/PyTorchLightning/pytorch-lightning/commit/3d5e71a7675c024560128f747c70c727fe95abd5#diff-37b14b9ef2f763796ae4fc2762c4ab6018bae62ea90f556aa960d7bc59695d9aR88

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -566,9 +568,13 @@ def select_precision_plugin(self) -> PrecisionPlugin:
...
-                    raise MisconfigurationException(
-                        "You have asked for native AMP on CPU, but AMP is only available on GPU."
-                    )
+                    if _TORCH_GREATER_EQUAL_1_10:
+                        return CPUNativeMixedPrecisionPlugin(self.precision)
+                    else:
+                        raise MisconfigurationException(
+                            "You have asked for native AMP on CPU, but AMP is only available on GPU for PyTorch 1.9 "
+                            "and lower. To use native AMP on CPU, install PyTorch 1.10 or later."
+                        )
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:577]:
```suggestion
                    raise MisconfigurationException(
                        "You have asked for native AMP on CPU, but AMP is only available on GPU for PyTorch 1.9 "
                        "and lower. To use native AMP on CPU, install PyTorch 1.10 or later."
                    )
```

File: pytorch_lightning/plugins/precision/cpu_native_amp.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class CPUNativeMixedPrecisionPlugin(NativeMixedPrecisionPlugin):
+    """
+    Plugin for CPU native mixed precision training with :mod:`torch.cpu.amp`.
+
+    Args:
+        precision: Whether to use torch.float16 (16) or torch.bfloat16 (bf16).
+            Currently only torch.bfloat16 is supported.
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/cpu_native_amp.py:29]: Is it worth mentioning 16 if it's not supported? Is it going to be supported before the 1.10 release?

File: pytorch_lightning/utilities/imports.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/utilities/imports.py:73]: great idea, will do!

File: pytorch_lightning/plugins/precision/cpu_native_amp.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py]: good point, don't think so! removed it

File: pytorch_lightning/plugins/precision/cpu_native_amp.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, precision: Union[int, str] = "bf16") -> None:
+        if not _TORCH_CPU_AMP_AVAILABLE:
+            raise MisconfigurationException(
+                "You have asked for native AMP on CPU, but AMP is only available on GPU for PyTorch 1.9 "
+                "and lower. To use native AMP on CPU, install PyTorch 1.10 or later."
+            )
+        super().__init__(precision)
+
+    def _select_precision_dtype(self, precision: Union[int, str] = 16) -> torch.dtype:
+        if not precision == "bf16":
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/cpu_native_amp.py:40]: Since this doesn't allow int anymore...

```suggestion
    def __init__(self, precision: str = "bf16") -> None:
        if not _TORCH_CPU_AMP_AVAILABLE:
            raise MisconfigurationException(
                "You have asked for native AMP on CPU, but AMP is only available on GPU for PyTorch 1.9 "
                "and lower. To use native AMP on CPU, install PyTorch 1.10 or later."
            )
        super().__init__(precision)

    def _select_precision_dtype(self, precision: str = "bf16") -> torch.dtype:
        if not precision == "bf16":
```

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Dope !

**Review by ananthsub — APPROVED** [review]
very nice!

File: pytorch_lightning/plugins/precision/cpu_native_amp.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    def _select_precision_dtype(self, precision: str = "bf16") -> torch.dtype:
+        if not precision == "bf16":
+            raise MisconfigurationException(
+                "CPU native amp only supports bfloat16. Please pass precision=bf16 to the Trainer."
+            )
+        return torch.bfloat16
+
+    def autocast_context_manager(self) -> torch.cuda.amp.autocast:
+        return torch.cpu.amp.autocast(fast_dtype=self._fast_dtype)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/cpu_native_amp.py:46]: we can't add this to NativeMixedPrecisionPlugin?

Comment (SeanNaren) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py:46]: We could totally by introducing a 'device' flag to the NativeMixedPrecisionPlugin, however that would involve some API work + enforcement of logic when on CPU (only bf16 supported). I felt like separating out the responsibilities since it's a completely different call would be reasonable!

the RFC is moving towards having it being all in one context manager: https://github.com/pytorch/pytorch/issues/55374 it seems like for now the API is split between cuda/cpu.

If anyone has any hard feelings here happy to swap between as it makes little difference to the user, primarily for lightning developers!

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py:46]: I just think that inheritance is not right here, because NativeMixedPrecisionPlugin already is an implementation. When you did the bfloat16 you could have also made a BfloatNativeMixedPrecision the same way, it just whould have required overriding more methods. 

since native means pytorch, and pytorch calls either into torch.cuda or torch.cpu, it is more reasonable to me to bring it under the same plugin. the only information we would need is whether we are on cpu or on cuda, but this we know at the time of instantiation in the accelerator connector, is my thinking correct?

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py:46]: I'm actually surprised @ananthsub did not request changes here as he is normally more concerned with this than me 😅

Comment (SeanNaren) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py:46]: Sure! I saw this coming initially but was curious if anyone would care/realise on this path. My initial thinking was cpu autocast involves completely different function calls, whereas bfloat was a dtype that can be passed to the plugin (hence adding the the same behaviour to the native precision plugin)!

More than happy to make the change to unblock this feature :)

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py:46]: what does @justusschock think?

Comment (ananthsub) [reply pytorch_lightning/plugins/precision/cpu_native_amp.py]: @awaelchli @SeanNaren tbh my hope is once we better spec out #7324 we can assess this streamlining then (@four4fish ).

**Comment by daniellepintz** [comment]
@SeanNaren I think this PR caused the following 5 tests to fail in our CI job (conda 3.7, 1.10)
```
FAILED tests/models/test_amp.py::test_amp_cpus[1-bf16-None] 
FAILED tests/models/test_amp.py::test_amp_cpus[1-bf16-ddp_spawn] 
FAILED tests/models/test_amp.py::test_amp_cpus[2-bf16-None] 
FAILED tests/models/test_amp.py::test_amp_cpus[2-bf16-ddp_spawn] 
FAILED tests/models/test_amp.py::test_cpu_model_with_amp 
```

The issue tracking this is #9280

Stack traces for the failing tests can be found in this gist: https://gist.github.com/daniellepintz/3d5a13b6264a832fc745f73460d2fa23 (from https://github.com/PyTorchLightning/pytorch-lightning/runs/3483032531)

Do you know what might be causing these failures?
